lib/buffer.js

```diff
@@ -539,11 +539,29 @@ Buffer.prototype.inspect = function inspect() {
   return '<' + this.constructor.name + ' ' + str + '>';
 };
 
+function assertInteger(val, name) {
+  if (typeof val !== 'undefined' &&
+      (typeof val !== 'number' ||
+       +val !== (val >> 0))) {
+    throw new TypeError(`"${name}" argument must be an integer`);
+  }
+}
 
-Buffer.prototype.compare = function compare(b) {
+Buffer.prototype.compare = function compare(b, start, end, thisStart, thisEnd) {
   if (!(b instanceof Buffer))
     throw new TypeError('Argument must be a Buffer');
 
+  assertInteger(start, 'targetStart');
+  assertInteger(thisEnd, 'sourceEnd');
+  assertInteger(end, 'targetEnd');
+  assertInteger(thisStart, 'sourceStart');
```

Comment:
Given what I said above, I'd suggest using checks closer to what's in `slowToString()`. We went through a lot of discussion making sure those checks/coersions to make sure those covered every case (ref ec836547).

That being said, I'd do it like so:
```js
if (thisEnd === undefined || thisEnd > this.length)
  thisEnd = this.length;
if (thisStart === undefined || thisStart < 0)
  thisStart = 0;
if (end === undefined || end > b.length)
  end = b.length;
if (start === undefined || start < 0)
  start = 0;

if (thisStart >= thisEnd && start >= end)
  return 0;
if (thisStart >= thisEnd)
  return -1;
if (start >= end)
return 1;

// Range comparisons above handle the case where end
// and thisEnd are < 0. So can do uint32 coercion without
// wrap around. Properly handles coercion of falsey/NaN
// values.
start >>>= 0;
end >>>= 0;
thisStart >>>= 0;
thisEnd >>>= 0;

return binding.compareOffset(this, b, start, thisStart, end, thisEnd);
```